devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
+                assert(len(provided_values) > 0)
+        
+                if len(provided_values) == 1:
+                    values[i] = provided_values[0]
+                else:
+                    values[i] = reduce(provided_values)
+        
+        # Second pass to evaluate any Unevaluated dependencies from the first pass
+        for k, v in values.items():
+            if isinstance(v, UnevaluatedDependency):
+                values[k] = v.evaluate(values)
+        return values
+
+    def _verify(self, values):
         verify = True
-        if value is None:
-            if self.value is not None:
-                return True
-
-            try:
-                parent_value = self.parent.value
-                if parent_value is None:
-                    return False
-            except AttributeError:
-                return False
-        # Make sure we're dealing with a 3-tuple.
-        # See docstring of infer_dimension_values_tuple for more info
-        value = infer_dimension_values_tuple(value, self.rtargs)
-        if hasattr(self, 'parent'):
-            parent_value = self.parent.value
-            if parent_value is not None and not enforce:
-                parent_value = infer_dimension_values_tuple(parent_value, self.rtargs)
-                value = tuple([self.reducer(i1, i2) for i1, i2 in zip(value,
-                                                                      parent_value)])
-            verify = verify and self.parent.verify(value)
-        if value == self.value:
-            return True
-
-        # Derived dimensions could be linked through constraints
-        # At this point, a constraint needs to be added that enforces
-        # dim_e - dim_s < SOME_MAX
-        # Also need a default constraint that dim_e > dim_s (or vice-versa)
-        verify = verify and all([a.verify(v, enforce=enforce) for a, v in
-                                 zip(self.rtargs, value)])
+        for i in values:
+            verify = verify and all(verify(i, x) for x in i.verified_by)
         return verify
 
+    @property
+    def dimensions(self):
+        return [x for x in self.parameters if isinstance(x, Dimension)]
 
-class ConstantArgProvider(ArgumentProvider):
 
-    """ Class used to decorate Constant Data objects with behaviour required for runtime
-        arguments.
+class ArgumentVisitor(GenericVisitor):
+    """ Visits types to return their runtime arguments
     """
-
-    @cached_property
-    def rtargs(self):
-        return (ScalarArgument(self.name, self, default_value=self.data),)
-
-
-class TensorFunctionArgProvider(ArgumentProvider):
-
-    """ Class used to decorate Symbolic Data objects with behaviour required for runtime
-        arguments.
+    def visit_SymbolicFunction(self, o):
+        return TensorArgument(o)
+    
+    def visit_DimensionParameter(self, o):
+        dependency = Dependency("gets_value_from", o)
```

Comment:
can we wrap these strings within a class? Have a look at ``IterationProperty`` and the various instances ``SEQUENTIAL, PARALLEL, ...`` for an example of what I mean

Reply:
Sure.